HW0/P4/P4.py

```diff
@@ -40,37 +44,108 @@
     # Read the observation array and plot it (Part 2)
     #####################
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    s_m = np.loadtxt('P4_measurements.txt', delimiter=',')
+    x_coords = 1.0 / rx * s_m[:,0]
+    y_coords = 1.0 / ry * s_m[:,1]
+    z_coords = 1.0 / rz * s_m[:,2]
+    ax.plot(x_coords, y_coords, z_coords,
+             '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
+    A = np.identity(6)
+    A[0:3,3:6] = np.identity(3) * dt
+    A[3:6,3:6] = np.identity(3) - c * dt * np.identity(3)
+    a = np.matrix([[0.0, 0.0, 0.0, 0.0, 0.0, g * dt]]).transpose()
 
     # Initial conditions for s0
+    s0 = np.matrix([[0, 0, 2, 15, 3.5, 4.0]]).transpose()
+
+
     # Compute the rest of sk using Eq (1)
+    s = np.zeros((6, K))
+    s[0:6,0] = s0.transpose()
+
+
+    # compute 
+    for i in range(1, K):
+        s_k = np.matrix(s[0:6, (i-1)]).transpose()
+        s_k1 = A * s_k + a
+        s[0:6, i] = s_k1.reshape((6))
+        #s[0:6, i] = (A * np.matrix(s[0:6, (i-1)]).transpose() + a).reshape((6)) # somehow python needs this reshaping command...
+
+    x_coords = s[0,:]
+    y_coords = s[1,:]
+    z_coords = s[2,:]
+
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    ax.plot(x_coords, y_coords, z_coords,
+             '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    B = np.matrix(np.diag([bx, by, bz, bvx, bvy, bvz]))
+    C = np.zeros((3, 6))
+    C[0:3,0:3] = np.diag([rx, ry, rz])
+    C = np.matrix(C)
+    # functions for the fourth part
+
+    # returns an intermediate guess which based on the propagation matrix and gravity
+    def predictS(s_k):
+        return A * s_k + a
+
+    # returns Sigma tilde
+    def predictSig(Sigma_k):
+        return (A * Sigma_k * A.T + B * B.T).I
+
+    # returns Sigma_{k+1}
+    def updateSig(Sigma_tilde):
+        return (Sigma_tilde + C.T * C).I
+
+    # use this function to retrieve s_{k+1} from s_k !
+    def updateS(Sigma_k1, Sigma_tilde, s_tilde, m_k1):
+        return Sigma_k1 * (Sigma_tilde * s_tilde + C.T * m_k1)
 
     # Initial conditions for s0 and Sigma0
+    Sigma_0 = 0.01 * np.identity(6)
```

Comment:
Try to avoid uppercase variable names, as these are generally reserved for class names in Python (see [PEP8](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables)).